### ISSUE TITLE
segmentation fault in lws_rx_flow_control when run cgi

### ISSUE THREAD

**Issue by zhiwenzheng** [warmcat/libwebsockets#1706]
> $c
libwebsockets.so.15`lws_rx_flow_control+0x16(0, 1, 8061bd8, fed161b5)
libwebsockets.so.15`lws_callback_http_dummy+0xfe(8124240, 10, 0, 0, 0, 4)
libwebsockets.so.15`user_callback_handle_rxflow+0x2f(fed161a4, 8124240, 10, 0, 0, 0)
libwebsockets.so.15`lws_callback_as_writeable+0xb1(8124240, fed5b4d6, 0, 8124240)
libwebsockets.so.15`lws_handle_POLLOUT_event+0x31d(8124240, 8063610, 8047828, fed26e06)
libwebsockets.so.15`rops_handle_POLLIN_h1+0x74(8061c68, 8124240, 8063610, fed1dfcb)
libwebsockets.so.15`lws_service_fd_tsi+0x22e(8061bd8, 8063610, 0, fed48d28)
libwebsockets.so.15`_lws_plat_service_forced_tsi+0x8e(8061bd8, 0, 3dc, 0)
libwebsockets.so.15`_lws_plat_service_tsi+0x36f(8061bd8, 77359400, 0, fed491c3)
libwebsockets.so.15`lws_plat_service+0x23(8061bd8, 3e8, fef70530, fed1e2db)
libwebsockets.so.15`lws_service+0x90(8061bd8, 3e8, 18c, fecdc380)
main+0xfc(8047c6c, fece43a8, 8047ca8, 80510b8, 1, 8047cd8)
_start_crt+0x97(1, 8047cd8, fefd1775, 0, 0, 0)
_start+0x1a(1, 8047dcc, 0, 8047dd9, 8047dff, 8047e10)

it is version 3.2.0 stable.
wsi->http.cgi->stdwsi[LWS_STDOUT] seems be NULL.
```
#ifdef LWS_WITH_CGI
311         if (wsi->reason_bf & (LWS_CB_REASON_AUX_BF__CGI_HEADERS |
312                       LWS_CB_REASON_AUX_BF__CGI)) {
313             n = lws_cgi_write_split_stdout_headers(wsi);
314             if (n < 0) {
315                 lwsl_debug("AUX_BF__CGI forcing close\n");
316                 return -1;
317             }       
318             if (!n)
319                 lws_rx_flow_control(
320                     wsi->http.cgi->stdwsi[LWS_STDOUT], 1);
```

**Comment by zhiwenzheng** [comment]
./flow.httpd 
[2019/10/05 18:13:49:6533] D: _realloc: size 6684: context
[2019/10/05 18:13:49:6534] D: _realloc: size 524288: fds table
[2019/10/05 18:13:49:6541] D: _realloc: size 262144: lws_lookup
[2019/10/05 18:13:49:6545] D: _realloc: size 472: event pipe wsi
[2019/10/05 18:13:49:6546] D: lws_role_transition: 8125ed8: wsistate 0x200, ops pipe
[2019/10/05 18:13:49:6547] D: event pipe fd 4
[2019/10/05 18:13:49:6547] D: __insert_wsi_socket_into_fds: 8125ed8: tsi=0, sock=4, pos-in-fds=0
[2019/10/05 18:13:49:6549] D: _realloc: size 544: create vhost
[2019/10/05 18:13:49:6549] D: _realloc: size 56: vhost-specific plugin table
[2019/10/05 18:13:49:6550] D: _realloc: size 12: same vh list
[2019/10/05 18:13:49:6551] D: _realloc: size 472: listen wsi
[2019/10/05 18:13:49:6552] D: _lws_vhost_init_server: lws_socket_bind says 81
[2019/10/05 18:13:49:6552] D: lws_role_transition: 8126928: wsistate 0x200, ops listen
[2019/10/05 18:13:49:6553] D: __insert_wsi_socket_into_fds: 8126928: tsi=0, sock=6, pos-in-fds=1
[2019/10/05 18:13:49:6554] D: _lws_state_transition: system: changed 1 'CONTEXT_CREATED' -> 2 'INITIALIZED'
[2019/10/05 18:13:49:6555] D: _lws_state_transition: system: changed 2 'INITIALIZED' -> 3 'TIME_VALID'
[2019/10/05 18:13:49:6555] D: _lws_state_transition: system: changed 3 'TIME_VALID' -> 4 'POLICY_VALID'
[2019/10/05 18:13:49:6556] D: _lws_state_transition: system: changed 4 'POLICY_VALID' -> 5 'OPERATIONAL'
[2019/10/05 18:13:49:6556] D: _lws_state_transition: system: changed 5 'OPERATIONAL' -> 4 'POLICY_VALID'
[2019/10/05 18:13:49:6557] D: _lws_state_transition: system: changed 4 'POLICY_VALID' -> 5 'OPERATIONAL'
[2019/10/05 18:15:01:0799] D: accepted new conn port 52408 on fd=7
[2019/10/05 18:15:01:0800] D: lws_get_idlest_tsi: 2 65535
[2019/10/05 18:15:01:0800] D: _realloc: size 472: new server wsi
[2019/10/05 18:15:01:0800] D: new wsi 8126b08 joining vhost default, tsi 0
[2019/10/05 18:15:01:0801] D: lwsi_set_state(8126b08, 0x20000200)
[2019/10/05 18:15:01:0801] D: lws_role_transition: 8126b08: wsistate 0x20000016, ops h1
[2019/10/05 18:15:01:0801] D: __lws_set_timeout: 8126b08: 20 secs, reason 3
[2019/10/05 18:15:01:0802] D: __insert_wsi_socket_into_fds: 8126b08: tsi=0, sock=7, pos-in-fds=2
[2019/10/05 18:15:01:0802] D: _realloc: size 952: ah struct
[2019/10/05 18:15:01:0802] D: _realloc: size 4096: ah data
[2019/10/05 18:15:01:0803] D: _lws_change_pollfd: wsi 8126b08: fd 7 events 1 -> 1
[2019/10/05 18:15:01:0804] D: __lws_set_timeout: 8126b08: 10 secs, reason 25
[2019/10/05 18:15:01:0804] D: Attached ah immediately
[2019/10/05 18:15:01:0809] D: rops_handle_POLLIN_h1: 8126b08: wsistate 0x20000016
[2019/10/05 18:15:01:0809] D: lws_read_h1: h1 path: wsi state 0x16
[2019/10/05 18:15:01:0810] D: lws_parse: unk header 26 'upgrade-insecure-requests:'
[2019/10/05 18:15:01:0811] D: lwsi_set_state(8126b08, 0x2000020a)
[2019/10/05 18:15:01:0811] D: lwsi_set_state(8126b08, 0x20000119)
[2019/10/05 18:15:01:0811] D: lws_handshake_server: wsi 8126b08: ah 8126ce8
[2019/10/05 18:15:01:0812] D: lws_ensure_user_space: 8126b08 protocol pss 0, user_space=0
[2019/10/05 18:15:01:0812] D: __lws_set_timeout: 8126b08: 20 secs, reason 10
[2019/10/05 18:15:01:0813] D: lws_http_action: cgi
[2019/10/05 18:15:01:0813] D: _realloc: size 264: new cgi
[2019/10/05 18:15:01:0814] D: _realloc: size 472: new wsi
[2019/10/05 18:15:01:0814] D: lws_role_transition: 81281c0: wsistate 0x119, ops cgi
[2019/10/05 18:15:01:0814] D: lws_cgi: cgi stdwsi 81281c0: pipe idx 0 -> fd 9 / 8
[2019/10/05 18:15:01:0815] D: _realloc: size 472: new wsi
[2019/10/05 18:15:01:0815] D: lws_role_transition: 81283a0: wsistate 0x119, ops cgi
[2019/10/05 18:15:01:0815] D: lws_cgi: cgi stdwsi 81283a0: pipe idx 1 -> fd 10 / 11
[2019/10/05 18:15:01:0816] D: _realloc: size 472: new wsi
[2019/10/05 18:15:01:0816] D: lws_role_transition: 8128580: wsistate 0x119, ops cgi
[2019/10/05 18:15:01:0816] D: lws_cgi: cgi stdwsi 8128580: pipe idx 2 -> fd 12 / 13
[2019/10/05 18:15:01:0816] D: __insert_wsi_socket_into_fds: 81281c0: tsi=0, sock=9, pos-in-fds=3
[2019/10/05 18:15:01:0817] D: __insert_wsi_socket_into_fds: 81283a0: tsi=0, sock=10, pos-in-fds=4
[2019/10/05 18:15:01:0817] D: __insert_wsi_socket_into_fds: 8128580: tsi=0, sock=12, pos-in-fds=5
[2019/10/05 18:15:01:0817] D: _lws_change_pollfd: wsi 81281c0: fd 9 events 1 -> 4
[2019/10/05 18:15:01:0818] D: _lws_change_pollfd: wsi 81283a0: fd 10 events 1 -> 1
[2019/10/05 18:15:01:0818] D: _lws_change_pollfd: wsi 8128580: fd 12 events 1 -> 1
[2019/10/05 18:15:01:0818] D: lws_cgi: fds in 9, out 10, err 12
[2019/10/05 18:15:01:0818] D: __lws_set_timeout: 8126b08: 5 secs, reason 14
[2019/10/05 18:15:01:0819] D: lws_cgi: adding cgi 81280b0 to list
[2019/10/05 18:15:01:0826] D: wsi->http.rx_content_length 0 0 0
[2019/10/05 18:15:01:0827] D: accepted new conn port 52410 on fd=8
[2019/10/05 18:15:01:0828] D: lws_get_idlest_tsi: 6 65535
[2019/10/05 18:15:01:0828] D: _realloc: size 472: new server wsi
[2019/10/05 18:15:01:0828] D: new wsi 8126ce8 joining vhost default, tsi 0
[2019/10/05 18:15:01:0829] D: lwsi_set_state(8126ce8, 0x20000200)
[2019/10/05 18:15:01:0829] D: lws_role_transition: 8126ce8: wsistate 0x20000016, ops h1
[2019/10/05 18:15:01:0829] D: __lws_set_timeout: 8126ce8: 20 secs, reason 3
[2019/10/05 18:15:01:0830] D: __insert_wsi_socket_into_fds: 8126ce8: tsi=0, sock=8, pos-in-fds=6
[2019/10/05 18:15:01:0830] D: _realloc: size 952: ah struct
[2019/10/05 18:15:01:0830] D: _realloc: size 4096: ah data
[2019/10/05 18:15:01:0831] D: _lws_change_pollfd: wsi 8126ce8: fd 8 events 1 -> 1
[2019/10/05 18:15:01:0831] D: __lws_set_timeout: 8126ce8: 10 secs, reason 25
[2019/10/05 18:15:01:0831] D: Attached ah immediately
[2019/10/05 18:15:01:0832] D: _lws_change_pollfd: wsi 81281c0: fd 9 events 4 -> 0
[2019/10/05 18:15:01:0832] D: CGI LWS_STDOUT 8126b08 wsistate 0x119
[2019/10/05 18:15:01:0880] D: CGI LWS_STDOUT 8126b08 wsistate 0x119
[2019/10/05 18:15:01:0880] D: _lws_change_pollfd: wsi 81283a0: fd 10 events 1 -> 0
[2019/10/05 18:15:01:0881] D: _lws_change_pollfd: wsi 8126b08: fd 7 events 1 -> 5
[2019/10/05 18:15:01:0881] D: _lws_change_pollfd: wsi 8126b08: fd 7 events 5 -> 1
[2019/10/05 18:15:01:0882] D: lws_handle_POLLOUT_event: 8126b08: non mux: wsistate 0x20000119, ops h1
[2019/10/05 18:15:01:0883] D: _realloc: size 2064: cgi hdr buf
[2019/10/05 18:15:01:0883] D: allocated cgi hdrs
[2019/10/05 18:15:01:0884] D: -- 0x53 S 0 0
[2019/10/05 18:15:01:0884] D: -- 0x74 t 0 0
[2019/10/05 18:15:01:0885] D: -- 0x61 a 0 0
[2019/10/05 18:15:01:0885] D: -- 0x74 t 0 0
[2019/10/05 18:15:01:0886] D: -- 0x75 u 0 0
[2019/10/05 18:15:01:0886] D: -- 0x73 s 0 0
[2019/10/05 18:15:01:0887] D: -- 0x3A : 0 0
[2019/10/05 18:15:01:0887] D: -- 0x20   0 0
[2019/10/05 18:15:01:0887] D: -- 0x34 4 0 0
[2019/10/05 18:15:01:0888] D: Status set to 4
[2019/10/05 18:15:01:0888] D: -- 0x30 0 0 0
[2019/10/05 18:15:01:0889] D: Status set to 40
[2019/10/05 18:15:01:0889] D: -- 0x31 1 0 0
[2019/10/05 18:15:01:0890] D: Status set to 401
[2019/10/05 18:15:01:0890] D: -- 0x20   0 0
[2019/10/05 18:15:01:0891] D: -- 0x55 U 0 0
[2019/10/05 18:15:01:0891] D: -- 0x6E n 0 0
[2019/10/05 18:15:01:0892] D: -- 0x61 a 0 0
[2019/10/05 18:15:01:0892] D: -- 0x75 u 0 0
[2019/10/05 18:15:01:0892] D: -- 0x74 t 0 0
[2019/10/05 18:15:01:0893] D: -- 0x68 h 0 0
[2019/10/05 18:15:01:0893] D: -- 0x6F o 0 0
[2019/10/05 18:15:01:0894] D: -- 0x72 r 0 0
[2019/10/05 18:15:01:0894] D: -- 0x69 i 0 0
[2019/10/05 18:15:01:0895] D: -- 0x7A z 0 0
[2019/10/05 18:15:01:0895] D: -- 0x65 e 0 0
[2019/10/05 18:15:01:0896] D: -- 0x64 d 0 0
 0 09/10/05 18:15:01:0896] D: -- 0x0D 
[2019/10/05 18:15:01:0897] D: -- 0x0A 
 0 2
 0 39/10/05 18:15:01:0897] D: -- 0x0D 
[2019/10/05 18:15:01:0898] D: -- 0x0A 
 0 4
[2019/10/05 18:15:01:0898] D: Content-Length: 0
[2019/10/05 18:15:01:0898] D: LHCS_RESPONSE: issuing response 401
[2019/10/05 18:15:01:0899] D: _lws_change_pollfd: using leave_pollout_active
[2019/10/05 18:15:01:0900] D: _lws_change_pollfd: wsi 81283a0: fd 10 events 0 -> 1
[2019/10/05 18:15:01:0900] D: _lws_change_pollfd: wsi 8126b08: fd 7 events 1 -> 5
[2019/10/05 18:15:01:0900] D: Session Socket 81281c0 (fd=9) dead
[2019/10/05 18:15:01:0901] D: 81281c0: Close and handled
[2019/10/05 18:15:01:0901] D: __lws_close_free_wsi: real just_kill_connection: 81281c0 (sockfd 9)
[2019/10/05 18:15:01:0901] D: _lws_change_pollfd: wsi 8126928: fd 6 events 1 -> 1
[2019/10/05 18:15:01:0902] D: lwsi_set_state(81281c0, 0x20)
[2019/10/05 18:15:01:0902] D: __lws_close_free_wsi_final: wsi 81281c0: fd 9
[2019/10/05 18:15:01:0902] D: __lws_free_wsi: 81281c0, remaining wsi 6, tsi fds count 6
[2019/10/05 18:15:01:0903] D: Session Socket 81283a0 (fd=10) dead
[2019/10/05 18:15:01:0903] D: 81283a0: Close and handled
[2019/10/05 18:15:01:0903] D: __lws_close_free_wsi: real just_kill_connection: 81283a0 (sockfd 10)
[2019/10/05 18:15:01:0903] D: _lws_change_pollfd: wsi 8126928: fd 6 events 1 -> 1
[2019/10/05 18:15:01:0904] D: lwsi_set_state(81283a0, 0x20)
[2019/10/05 18:15:01:0904] D: __lws_close_free_wsi_final: wsi 81283a0: fd 10
[2019/10/05 18:15:01:0904] D: __lws_free_wsi: 81283a0, remaining wsi 5, tsi fds count 5
[2019/10/05 18:15:01:0905] D: Session Socket 8128580 (fd=12) dead
[2019/10/05 18:15:01:0905] D: 8128580: Close and handled
[2019/10/05 18:15:01:0905] D: lws_cgi_remove_and_kill: remove cgi 81280b0 from list
[2019/10/05 18:15:01:0905] D: close: freed cgi headers
[2019/10/05 18:15:01:0906] D: lws_cgi_kill: 8126b08
[2019/10/05 18:15:01:0906] D: lws_cgi_kill: PID 100514 reaped
[2019/10/05 18:15:01:0907] D: LWS_CALLBACK_CGI_TERMINATED: 0 0
[2019/10/05 18:15:01:0907] D: LWS_CALLBACK_CGI_TERMINATED: ending
[2019/10/05 18:15:01:0907] D: _lws_change_pollfd: wsi 8126b08: fd 7 events 5 -> 5
[2019/10/05 18:15:01:0907] D: __lws_set_timeout: 8126b08: 3 secs, reason 14
[2019/10/05 18:15:01:0908] D: __lws_close_free_wsi: real just_kill_connection: 8128580 (sockfd 12)
[2019/10/05 18:15:01:0908] D: _lws_change_pollfd: wsi 8126928: fd 6 events 1 -> 1
[2019/10/05 18:15:01:0908] D: lwsi_set_state(8128580, 0x20)
[2019/10/05 18:15:01:0908] D: __lws_close_free_wsi_final: wsi 8128580: fd 12
[2019/10/05 18:15:01:0909] D: __lws_free_wsi: 8128580, remaining wsi 4, tsi fds count 4
[2019/10/05 18:15:01:0909] D: _lws_change_pollfd: wsi 8126b08: fd 7 events 5 -> 1
[2019/10/05 18:15:01:0909] D: lws_handle_POLLOUT_event: 8126b08: non mux: wsistate 0x20000119, ops h1
[2019/10/05 18:15:01:0910] D: LHCS_DUMP_HEADERS: 28
[2019/10/05 18:15:01:0910] D: freed cgi headers
Segmentation Fault (core dumped)

**Comment by lws-team** [comment]
I pushed a patch on master and v3.2-stable that should address it... but I'm not sure what it means that the cgi stdout is NULL at that time... it was in process of draining / closing when the WRITABLE came, or somehow it came very early before it was set up?  Or it failed?

**Comment by zhiwenzheng** [comment]
The patch fixed the fault. But the cgi server only send the response header, the body message is not send.

[2019/10/06 16:43:50:2137] I: Method: 'GET' (0), request for '/man'
[2019/10/06 16:43:50:2137] D: lws_ensure_user_space: 8126b08 protocol pss 0, user_space=0
[2019/10/06 16:43:50:2137] D: __lws_set_timeout: 8126b08: 20 secs, reason 10
[2019/10/06 16:43:50:2138] D: lws_http_action: cgi
[2019/10/06 16:43:50:2138] D: _realloc: size 264: new cgi
[2019/10/06 16:43:50:2139] D: _realloc: size 472: new wsi
[2019/10/06 16:43:50:2139] D: lws_role_transition: 8129798: wsistate 0x119, ops cgi
[2019/10/06 16:43:50:2140] I: lws_vhost_bind_wsi: vh default: count_bound_wsi 4
[2019/10/06 16:43:50:2140] D: lws_cgi: cgi stdwsi 8129798: pipe idx 0 -> fd 10 / 8
[2019/10/06 16:43:50:2140] D: _realloc: size 472: new wsi
[2019/10/06 16:43:50:2140] D: lws_role_transition: 8129978: wsistate 0x119, ops cgi
[2019/10/06 16:43:50:2141] I: lws_vhost_bind_wsi: vh default: count_bound_wsi 5
[2019/10/06 16:43:50:2141] D: lws_cgi: cgi stdwsi 8129978: pipe idx 1 -> fd 11 / 12
[2019/10/06 16:43:50:2141] D: _realloc: size 472: new wsi
[2019/10/06 16:43:50:2142] D: lws_role_transition: 8129b58: wsistate 0x119, ops cgi
[2019/10/06 16:43:50:2142] I: lws_vhost_bind_wsi: vh default: count_bound_wsi 6
[2019/10/06 16:43:50:2142] D: lws_cgi: cgi stdwsi 8129b58: pipe idx 2 -> fd 13 / 14
[2019/10/06 16:43:50:2143] D: __insert_wsi_socket_into_fds: 8129798: tsi=0, sock=10, pos-in-fds=4
[2019/10/06 16:43:50:2143] D: __insert_wsi_socket_into_fds: 8129978: tsi=0, sock=11, pos-in-fds=5
[2019/10/06 16:43:50:2143] D: __insert_wsi_socket_into_fds: 8129b58: tsi=0, sock=13, pos-in-fds=6
[2019/10/06 16:43:50:2143] D: _lws_change_pollfd: wsi 8129798: fd 10 events 1 -> 4
[2019/10/06 16:43:50:2144] D: _lws_change_pollfd: wsi 8129978: fd 11 events 1 -> 1
[2019/10/06 16:43:50:2144] D: _lws_change_pollfd: wsi 8129b58: fd 13 events 1 -> 1
[2019/10/06 16:43:50:2144] D: lws_cgi: fds in 10, out 11, err 13
[2019/10/06 16:43:50:2145] D: __lws_set_timeout: 8126b08: 30 secs, reason 14
[2019/10/06 16:43:50:2145] D: lws_cgi: adding cgi 8129688 to list
[2019/10/06 16:43:50:2145] D:     REQUEST_METHOD=GET
[2019/10/06 16:43:50:2145] D:     QUERY_STRING=req=%7B%22o%22%3A%22ipaddr%22%2C%22m%22%3A%22list%22%2C%22p%22%3A%22%22%7D
[2019/10/06 16:43:50:2146] D:     REQUEST_URI=/man
[2019/10/06 16:43:50:2146] D:     PATH_INFO=
[2019/10/06 16:43:50:2146] D:     HTTP_REFERER=http://h55:81/
[2019/10/06 16:43:50:2146] D:     HTTP_HOST=h55:81
[2019/10/06 16:43:50:2147] D:     HTTP_COOKIE=User=TAANLw==.lHp0tU7QDpRNQisAJa4CwA==.V4wbMoiwOQzGtnmjzSneBA==
[2019/10/06 16:43:50:2147] D:     HTTP_USER_AGENT=Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/76.0.3809.100 Safari/537.36
[2019/10/06 16:43:50:2147] D:     HTTP_ACCEPT=application/json, text/plain, */*
[2019/10/06 16:43:50:2147] D:     HTTP_ACCEPT_ENCODING=gzip, deflate
[2019/10/06 16:43:50:2148] D:     PATH=/bin:/usr/bin:/usr/local/bin:/var/www/cgi-bin
[2019/10/06 16:43:50:2148] D:     SCRIPT_NAME=/var/www/man
[2019/10/06 16:43:50:2148] D:     SERVER_SOFTWARE=libwebsockets
[2019/10/06 16:43:50:2148] I: __lws_header_table_detach: wsi 8126b08: ah 8126ce8 (tsi=0, count = 2)
[2019/10/06 16:43:50:2149] I: __lws_header_table_detach: nobody usable waiting
[2019/10/06 16:43:50:2149] I: _lws_destroy_ah: freed ah 8126ce8 : pool length 1
[2019/10/06 16:43:50:2149] I: __lws_header_table_detach: wsi 8126b08: ah 8126ce8 (tsi=0, count = 1)
[2019/10/06 16:43:50:2157] I: lws_cgi: cgi 8129688 spawned PID 100531
[2019/10/06 16:43:50:2157] D: wsi->http.rx_content_length 0 0 0
[2019/10/06 16:43:50:2158] D: _lws_change_pollfd: wsi 8129798: fd 10 events 4 -> 0
[2019/10/06 16:43:50:2158] D: CGI LWS_STDOUT 8126b08 wsistate 0x119
[2019/10/06 16:43:50:2166] D: rops_handle_POLLIN_h1: 81280e0: wsistate 0x20000016
[2019/10/06 16:43:50:2166] I: lws_buflist_aware_read: wsi 81280e0: lws_h1_server_socket_service: ssl_capable_read 351 (prior 0)
[2019/10/06 16:43:50:2166] D: lws_read_h1: h1 path: wsi state 0x16
[2019/10/06 16:43:50:2167] I: lws_handshake_server: parsed count 351
[2019/10/06 16:43:50:2167] I: lws_select_vhost: vhost match to default based on port 81
[2019/10/06 16:43:50:2168] D: lwsi_set_state(81280e0, 0x2000020a)
[2019/10/06 16:43:50:2168] I: lws_handshake_server: 81280e0: No upgrade
[2019/10/06 16:43:50:2168] D: lwsi_set_state(81280e0, 0x20000119)
[2019/10/06 16:43:50:2168] D: lws_handshake_server: wsi 81280e0: ah 81282c0
[2019/10/06 16:43:50:2169] I: Method: 'GET' (0), request for '/js/app.js.map'
[2019/10/06 16:43:50:2169] D: lws_ensure_user_space: 81280e0 protocol pss 0, user_space=0
[2019/10/06 16:43:50:2169] D: __lws_set_timeout: 81280e0: 20 secs, reason 10
[2019/10/06 16:43:50:2169] I: client indicates GZIP is acceptable
[2019/10/06 16:43:50:2170] I: lws_http_serve: Unable to open '/usr/lib/flow/http/js/app.js.map': errno 2
[2019/10/06 16:43:50:2170] I: lws_issue_raw: ssl_capable_write (245) says 245
[2019/10/06 16:43:50:2171] I: lws_http_transaction_completed: wsi 81280e0
[2019/10/06 16:43:50:2171] D: lws_ensure_user_space: 81280e0 protocol pss 0, user_space=0
[2019/10/06 16:43:50:2171] D: lws_http_transaction_completed: 81280e0: setting DEF_ACT from 0x20000119
[2019/10/06 16:43:50:2172] D: lwsi_set_state(81280e0, 0x20000110)
[2019/10/06 16:43:50:2172] D: __lws_set_timeout: 81280e0: 5 secs, reason 15
[2019/10/06 16:43:50:2172] D: lws_http_transaction_completed: 81280e0: nothing in buflist, detaching ah
[2019/10/06 16:43:50:2172] I: __lws_header_table_detach: wsi 81280e0: ah 81282c0 (tsi=0, count = 1)
[2019/10/06 16:43:50:2173] I: __lws_header_table_detach: nobody usable waiting
[2019/10/06 16:43:50:2173] I: _lws_destroy_ah: freed ah 81282c0 : pool length 0
[2019/10/06 16:43:50:2173] I: __lws_header_table_detach: wsi 81280e0: ah 81282c0 (tsi=0, count = 0)
[2019/10/06 16:43:50:2173] D: lws_http_transaction_completed: 81280e0: keep-alive await new transaction (state 0x20000110)
[2019/10/06 16:43:50:2174] D: _lws_change_pollfd: wsi 81280e0: fd 9 events 1 -> 5
[2019/10/06 16:43:50:2174] D: wsi->http.rx_content_length 0 0 0
[2019/10/06 16:43:50:2174] D: rops_handle_POLLIN_h1: 81280e0: wsistate 0x20000110
[2019/10/06 16:43:50:2175] D: _lws_change_pollfd: wsi 81280e0: fd 9 events 5 -> 1
[2019/10/06 16:43:50:2175] D: lws_h1_server_socket_service: LRS_DEFERRING_ACTION now writable
[2019/10/06 16:43:50:2175] D: lwsi_set_state(81280e0, 0x20000119)
[2019/10/06 16:43:50:2175] D: _lws_change_pollfd: wsi 81280e0: fd 9 events 1 -> 1
[2019/10/06 16:43:50:2326] D: rops_handle_POLLIN_h1: 81280e0: wsistate 0x20000119
[2019/10/06 16:43:50:2327] I: lws_header_table_attach: wsi 81280e0: ah 0 (tsi 0, count = 0) in
[2019/10/06 16:43:50:2327] D: _realloc: size 952: ah struct
[2019/10/06 16:43:50:2327] D: _realloc: size 4096: ah data
[2019/10/06 16:43:50:2328] I: _lws_create_ah: created ah 81282c0 (size 4096): pool length 1
[2019/10/06 16:43:50:2328] D: _lws_change_pollfd: wsi 81280e0: fd 9 events 1 -> 1
[2019/10/06 16:43:50:2328] I: lws_header_table_attach: did attach wsi 81280e0: ah 81282c0: count 1 (on exit)
[2019/10/06 16:43:50:2328] D: __lws_set_timeout: 81280e0: 10 secs, reason 25
[2019/10/06 16:43:50:2329] I: lws_buflist_aware_read: wsi 81280e0: lws_h1_server_socket_service: ssl_capable_read 441 (prior 0)
[2019/10/06 16:43:50:2329] D: lws_read_h1: h1 path: wsi state 0x119
[2019/10/06 16:43:50:2330] I: lws_handshake_server: parsed count 441
[2019/10/06 16:43:50:2330] I: lws_select_vhost: vhost match to default based on port 81
[2019/10/06 16:43:50:2330] D: lwsi_set_state(81280e0, 0x2000020a)
[2019/10/06 16:43:50:2331] I: lws_handshake_server: 81280e0: No upgrade
[2019/10/06 16:43:50:2331] D: lwsi_set_state(81280e0, 0x20000119)
[2019/10/06 16:43:50:2331] D: lws_handshake_server: wsi 81280e0: ah 81282c0
[2019/10/06 16:43:50:2331] I: Method: 'GET' (0), request for '/favicon.ico'
[2019/10/06 16:43:50:2332] D: lws_ensure_user_space: 81280e0 protocol pss 0, user_space=0
[2019/10/06 16:43:50:2332] D: __lws_set_timeout: 81280e0: 20 secs, reason 10
[2019/10/06 16:43:50:2332] I: client indicates GZIP is acceptable
[2019/10/06 16:43:50:2333] I: lws_get_mimetype: match to server mimetype: image/x-icon
[2019/10/06 16:43:50:2333] I: lws_add_http_header_content_length: wsi 81280e0: tx_content_length/remain 4286
[2019/10/06 16:43:50:2334] I: lws_issue_raw: ssl_capable_write (118) says 118
[2019/10/06 16:43:50:2334] D: lwsi_set_state(81280e0, 0x20000015)
[2019/10/06 16:43:50:2334] D: _lws_change_pollfd: wsi 81280e0: fd 9 events 1 -> 5
[2019/10/06 16:43:50:2335] D: rops_handle_POLLIN_h1: 81280e0: wsistate 0x20000015
[2019/10/06 16:43:50:2335] D: _lws_change_pollfd: wsi 81280e0: fd 9 events 5 -> 1
[2019/10/06 16:43:50:2335] D: wsi->http2_substream 0
[2019/10/06 16:43:50:2336] D: _lws_plat_file_read: read 4087 of req 4087, pos 4087, len 4286
[2019/10/06 16:43:50:2336] D: lws_serve_http_file_fragment: sending 4087
[2019/10/06 16:43:50:2336] D: __lws_set_timeout: 81280e0: 20 secs, reason 10
[2019/10/06 16:43:50:2337] I: lws_issue_raw: ssl_capable_write (4087) says 4087
[2019/10/06 16:43:50:2337] D: _lws_plat_file_read: read 199 of req 4087, pos 4286, len 4286
[2019/10/06 16:43:50:2338] D: lws_serve_http_file_fragment: sending 199
[2019/10/06 16:43:50:2339] D: __lws_set_timeout: 81280e0: 20 secs, reason 10
[2019/10/06 16:43:50:2340] I: lws_issue_raw: ssl_capable_write (199) says 199
[2019/10/06 16:43:50:2340] D: lwsi_set_state(81280e0, 0x20000119)
[2019/10/06 16:43:50:2340] D: file completed
[2019/10/06 16:43:50:2341] I: lws_http_transaction_completed: wsi 81280e0
[2019/10/06 16:43:50:2341] D: lws_ensure_user_space: 81280e0 protocol pss 0, user_space=0
[2019/10/06 16:43:50:2342] D: lws_http_transaction_completed: 81280e0: setting DEF_ACT from 0x20000119
[2019/10/06 16:43:50:2343] D: lwsi_set_state(81280e0, 0x20000110)
[2019/10/06 16:43:50:2343] D: __lws_set_timeout: 81280e0: 5 secs, reason 15
[2019/10/06 16:43:50:2344] D: lws_http_transaction_completed: 81280e0: nothing in buflist, detaching ah
[2019/10/06 16:43:50:2344] I: __lws_header_table_detach: wsi 81280e0: ah 81282c0 (tsi=0, count = 1)
[2019/10/06 16:43:50:2344] I: __lws_header_table_detach: nobody usable waiting
[2019/10/06 16:43:50:2345] I: _lws_destroy_ah: freed ah 81282c0 : pool length 0
[2019/10/06 16:43:50:2345] I: __lws_header_table_detach: wsi 81280e0: ah 81282c0 (tsi=0, count = 0)
[2019/10/06 16:43:50:2345] D: lws_http_transaction_completed: 81280e0: keep-alive await new transaction (state 0x20000110)
[2019/10/06 16:43:50:2345] D: _lws_change_pollfd: wsi 81280e0: fd 9 events 1 -> 5
[2019/10/06 16:43:50:2346] D: rops_handle_POLLIN_h1: 81280e0: wsistate 0x20000110
[2019/10/06 16:43:50:2346] D: _lws_change_pollfd: wsi 81280e0: fd 9 events 5 -> 1
[2019/10/06 16:43:50:2346] D: lws_h1_server_socket_service: LRS_DEFERRING_ACTION now writable
[2019/10/06 16:43:50:2347] D: lwsi_set_state(81280e0, 0x20000119)
[2019/10/06 16:43:50:2347] D: _lws_change_pollfd: wsi 81280e0: fd 9 events 1 -> 1
[2019/10/06 16:43:50:2396] D: CGI LWS_STDOUT 8126b08 wsistate 0x119
[2019/10/06 16:43:50:2397] I: lws_rx_flow_control: 8129978 0x0
[2019/10/06 16:43:50:2397] I: lws_rx_flow_control: 8129978: bitmap 0x1: en 0x4001, ch 0x2
[2019/10/06 16:43:50:2398] I: rxflow: wsi 8129978 change_to 0
[2019/10/06 16:43:50:2398] D: _lws_change_pollfd: wsi 8129978: fd 11 events 1 -> 0
[2019/10/06 16:43:50:2399] D: _lws_change_pollfd: wsi 8126b08: fd 7 events 1 -> 5
[2019/10/06 16:43:50:2399] D: _lws_change_pollfd: wsi 8126b08: fd 7 events 5 -> 1
[2019/10/06 16:43:50:2400] D: lws_handle_POLLOUT_event: 8126b08: non mux: wsistate 0x20000119, ops h1
[2019/10/06 16:43:50:2400] D: _realloc: size 2064: cgi hdr buf
[2019/10/06 16:43:50:2401] D: allocated cgi hdrs
[2019/10/06 16:43:50:2401] D: -- 0x43 C 0 0
[2019/10/06 16:43:50:2402] D: -- 0x6F o 0 0
[2019/10/06 16:43:50:2402] D: -- 0x6E n 0 0
[2019/10/06 16:43:50:2403] D: -- 0x74 t 0 0
[2019/10/06 16:43:50:2403] D: -- 0x65 e 0 0
[2019/10/06 16:43:50:2403] D: -- 0x6E n 0 0
[2019/10/06 16:43:50:2404] D: -- 0x74 t 0 0
[2019/10/06 16:43:50:2404] D: -- 0x2D - 0 0
[2019/10/06 16:43:50:2405] D: -- 0x74 t 0 0
[2019/10/06 16:43:50:2405] D: -- 0x79 y 0 0
[2019/10/06 16:43:50:2405] D: -- 0x70 p 0 0
[2019/10/06 16:43:50:2406] D: -- 0x65 e 0 0
[2019/10/06 16:43:50:2406] D: -- 0x3A : 0 0
[2019/10/06 16:43:50:2407] D: -- 0x20   0 0
[2019/10/06 16:43:50:2407] D: -- 0x61 a 0 0
[2019/10/06 16:43:50:2407] D: -- 0x70 p 0 0
[2019/10/06 16:43:50:2408] D: -- 0x70 p 0 0
[2019/10/06 16:43:50:2408] D: -- 0x6C l 0 0
[2019/10/06 16:43:50:2409] D: -- 0x69 i 1 0
[2019/10/06 16:43:50:2409] D: -- 0x63 c 0 0
[2019/10/06 16:43:50:2410] D: -- 0x61 a 0 0
[2019/10/06 16:43:50:2410] D: -- 0x74 t 0 0
[2019/10/06 16:43:50:2411] D: -- 0x69 i 0 0
[2019/10/06 16:43:50:2411] D: -- 0x6F o 0 0
[2019/10/06 16:43:50:2411] D: -- 0x6E n 0 0
[2019/10/06 16:43:50:2411] D: -- 0x2F / 0 0
[2019/10/06 16:43:50:2412] D: -- 0x6A j 0 0
[2019/10/06 16:43:50:2412] D: -- 0x73 s 0 0
[2019/10/06 16:43:50:2412] D: -- 0x6F o 0 0
[2019/10/06 16:43:50:2413] D: -- 0x6E n 0 0
 0 09/10/06 16:43:50:2413] D: -- 0x0D 
[2019/10/06 16:43:50:2413] D: -- 0x0A 
 0 2
[2019/10/06 16:43:50:2413] D: -- 0x43 C 0 3
[2019/10/06 16:43:50:2414] D: -- 0x6F o 0 0
[2019/10/06 16:43:50:2414] D: -- 0x6E n 0 0
[2019/10/06 16:43:50:2414] D: -- 0x74 t 0 0
[2019/10/06 16:43:50:2415] D: -- 0x65 e 0 0
[2019/10/06 16:43:50:2415] D: -- 0x6E n 0 0
[2019/10/06 16:43:50:2415] D: -- 0x74 t 0 0
[2019/10/06 16:43:50:2416] D: -- 0x2D - 0 0
[2019/10/06 16:43:50:2416] D: -- 0x6C l 0 0
[2019/10/06 16:43:50:2416] D: -- 0x65 e 1 0
[2019/10/06 16:43:50:2417] D: -- 0x6E n 0 0
[2019/10/06 16:43:50:2417] D: -- 0x67 g 0 0
[2019/10/06 16:43:50:2417] D: -- 0x74 t 0 0
[2019/10/06 16:43:50:2418] D: -- 0x68 h 0 0
[2019/10/06 16:43:50:2418] D: -- 0x3A : 0 0
[2019/10/06 16:43:50:2418] D: -- 0x20   0 0
[2019/10/06 16:43:50:2418] D: -- 0x32 2 0 0
[2019/10/06 16:43:50:2419] D: -- 0x32 2 0 0
[2019/10/06 16:43:50:2420] D: -- 0x32 2 0 0
 0 09/10/06 16:43:50:2420] D: -- 0x0D 
[2019/10/06 16:43:50:2420] D: -- 0x0A 
 0 2
 0 39/10/06 16:43:50:2420] D: -- 0x0D 
[2019/10/06 16:43:50:2421] D: -- 0x0A 
 0 4
[2019/10/06 16:43:50:2421] D: Content-Length: 222
[2019/10/06 16:43:50:2421] D: LHCS_RESPONSE: issuing response 200
[2019/10/06 16:43:50:2422] I: lws_issue_raw: ssl_capable_write (36) says 36
[2019/10/06 16:43:50:2422] D: _lws_change_pollfd: using leave_pollout_active
[2019/10/06 16:43:50:2422] I: lws_rx_flow_control: 8129978 0x1
[2019/10/06 16:43:50:2422] I: lws_rx_flow_control: 8129978: bitmap 0x0: en 0x6001, ch 0x3
[2019/10/06 16:43:50:2423] I: rxflow: wsi 8129978 change_to 1
[2019/10/06 16:43:50:2423] I: __lws_rx_flow_control: reenable POLLIN
[2019/10/06 16:43:50:2423] D: _lws_change_pollfd: wsi 8129978: fd 11 events 0 -> 1
[2019/10/06 16:43:50:2424] D: _lws_change_pollfd: wsi 8126b08: fd 7 events 1 -> 5
[2019/10/06 16:43:50:2424] D: Session Socket 8129798 (fd=10) dead
[2019/10/06 16:43:50:2424] D: 8129798: Close and handled
[2019/10/06 16:43:50:2424] I: __lws_close_free_wsi: 8129798: caller: close_and_handled
[2019/10/06 16:43:50:2425] D: __lws_close_free_wsi: real just_kill_connection: 8129798 (sockfd 10)
[2019/10/06 16:43:50:2425] D: _lws_change_pollfd: wsi 8126928: fd 6 events 1 -> 1
[2019/10/06 16:43:50:2425] D: lwsi_set_state(8129798, 0x20)
[2019/10/06 16:43:50:2426] I: lws_remove_child_from_any_parent: detach 8129798 from parent 8126b08
[2019/10/06 16:43:50:2426] D: __lws_close_free_wsi_final: wsi 8129798: fd 10
[2019/10/06 16:43:50:2426] I: lws_vhost_unbind_wsi: vh default: count_bound_wsi 5
[2019/10/06 16:43:50:2427] D: __lws_free_wsi: 8129798, remaining wsi 6, tsi fds count 6
[2019/10/06 16:43:50:2427] D: Session Socket 8129b58 (fd=13) dead
[2019/10/06 16:43:50:2427] D: 8129b58: Close and handled
[2019/10/06 16:43:50:2428] I: __lws_close_free_wsi: 8129b58: caller: close_and_handled
[2019/10/06 16:43:50:2428] D: __lws_close_free_wsi: real just_kill_connection: 8129b58 (sockfd 13)
[2019/10/06 16:43:50:2428] D: _lws_change_pollfd: wsi 8126928: fd 6 events 1 -> 1
[2019/10/06 16:43:50:2428] D: lwsi_set_state(8129b58, 0x20)
[2019/10/06 16:43:50:2429] I: lws_remove_child_from_any_parent: detach 8129b58 from parent 8126b08
[2019/10/06 16:43:50:2429] D: __lws_close_free_wsi_final: wsi 8129b58: fd 13
[2019/10/06 16:43:50:2429] I: lws_vhost_unbind_wsi: vh default: count_bound_wsi 4
[2019/10/06 16:43:50:2430] D: __lws_free_wsi: 8129b58, remaining wsi 5, tsi fds count 5
[2019/10/06 16:43:50:2430] D: Session Socket 8129978 (fd=11) dead
[2019/10/06 16:43:50:2430] D: 8129978: Close and handled
[2019/10/06 16:43:50:2430] I: __lws_close_free_wsi: 8129978: caller: close_and_handled
[2019/10/06 16:43:50:2431] D: lws_cgi_remove_and_kill: remove cgi 8129688 from list
[2019/10/06 16:43:50:2431] D: close: freed cgi headers
[2019/10/06 16:43:50:2431] D: lws_cgi_kill: 8126b08
[2019/10/06 16:43:50:2432] D: lws_cgi_kill: PID 100531 reaped
[2019/10/06 16:43:50:2432] D: LWS_CALLBACK_CGI_TERMINATED: 0 222
[2019/10/06 16:43:50:2432] I: lws_http_transaction_completed: wsi 8126b08
[2019/10/06 16:43:50:2433] D: lws_ensure_user_space: 8126b08 protocol pss 0, user_space=0
[2019/10/06 16:43:50:2433] D: lws_http_transaction_completed: 8126b08: setting DEF_ACT from 0x20000119
[2019/10/06 16:43:50:2433] D: lwsi_set_state(8126b08, 0x20000110)
[2019/10/06 16:43:50:2433] D: __lws_set_timeout: 8126b08: 5 secs, reason 15
[2019/10/06 16:43:50:2434] D: lws_http_transaction_completed: 8126b08: keep-alive await new transaction (state 0x20000110)
[2019/10/06 16:43:50:2434] D: _lws_change_pollfd: wsi 8126b08: fd 7 events 5 -> 5
[2019/10/06 16:43:50:2434] D: __lws_close_free_wsi: real just_kill_connection: 8129978 (sockfd 11)
[2019/10/06 16:43:50:2434] D: _lws_change_pollfd: wsi 8126928: fd 6 events 1 -> 1
[2019/10/06 16:43:50:2435] D: lwsi_set_state(8129978, 0x20)
[2019/10/06 16:43:50:2435] I: lws_remove_child_from_any_parent: detach 8129978 from parent 8126b08
[2019/10/06 16:43:50:2435] D: __lws_close_free_wsi_final: wsi 8129978: fd 11
[2019/10/06 16:43:50:2436] I: lws_vhost_unbind_wsi: vh default: count_bound_wsi 3
[2019/10/06 16:43:50:2436] D: __lws_free_wsi: 8129978, remaining wsi 4, tsi fds count 4
[2019/10/06 16:43:50:2437] D: _lws_change_pollfd: wsi 8126b08: fd 7 events 5 -> 1
[2019/10/06 16:43:50:2437] D: lws_handle_POLLOUT_event: 8126b08: non mux: wsistate 0x20000110, ops h1
[2019/10/06 16:43:50:2437] D: LHCS_DUMP_HEADERS: 55
[2019/10/06 16:43:50:2438] I: lws_issue_raw: ssl_capable_write (55) says 55
[2019/10/06 16:43:50:2438] D: freed cgi headers
[2019/10/06 16:43:55:2343] I: wsi 81280e0: TIMEDOUT WAITING on 15 (did hdr 0, ah 0, wl 0)
[2019/10/06 16:43:55:2344] I: __lws_close_free_wsi: 81280e0: caller: timeout
[2019/10/06 16:43:55:2344] I: __lws_close_free_wsi: shutdown conn: 81280e0 (sk 9, state 0x119)
[2019/10/06 16:43:55:2345] D: __lws_close_free_wsi: real just_kill_connection: 81280e0 (sockfd 9)
[2019/10/06 16:43:55:2346] D: _lws_change_pollfd: wsi 8126928: fd 6 events 1 -> 1
[2019/10/06 16:43:55:2346] D: lwsi_set_state(81280e0, 0x20000020)
[2019/10/06 16:43:55:2346] D: __lws_close_free_wsi_final: wsi 81280e0: fd 9
[2019/10/06 16:43:55:2347] I: rops_destroy_role_h1: ah det due to close
[2019/10/06 16:43:55:2347] I: lws_vhost_unbind_wsi: vh default: count_bound_wsi 2
[2019/10/06 16:43:55:2348] D: __lws_free_wsi: 81280e0, remaining wsi 3, tsi fds count 3
[2019/10/06 16:43:55:2438] I: wsi 8126b08: TIMEDOUT WAITING on 15 (did hdr 0, ah 0, wl 0)
[2019/10/06 16:43:55:2439] N: CGI timeout: 
[2019/10/06 16:43:55:2439] I: __lws_close_free_wsi: 8126b08: caller: timeout
[2019/10/06 16:43:55:2439] D: lws_cgi_remove_and_kill: remove cgi 8129688 from list
[2019/10/06 16:43:55:2440] D: lws_cgi_kill: 8126b08
[2019/10/06 16:43:55:2440] I: __lws_close_free_wsi: shutdown conn: 8126b08 (sk 7, state 0x110)
[2019/10/06 16:43:55:2440] D: __lws_close_free_wsi: real just_kill_connection: 8126b08 (sockfd 7)
[2019/10/06 16:43:55:2441] D: _lws_change_pollfd: wsi 8126928: fd 6 events 1 -> 1
[2019/10/06 16:43:55:2441] D: lwsi_set_state(8126b08, 0x20000020)
[2019/10/06 16:43:55:2441] D: __lws_close_free_wsi_final: wsi 8126b08: fd 7
[2019/10/06 16:43:55:2442] I: rops_destroy_role_h1: ah det due to close
[2019/10/06 16:43:55:2442] I: lws_vhost_unbind_wsi: vh default: count_bound_wsi 1
[2019/10/06 16:43:55:2442] D: __lws_free_wsi: 8126b08, remaining wsi 2, tsi fds count 2

I looked into the source code, but do not understand it well enough, it seems that,
lws_cgi_remove_and_kill() get called, when payload has not been processed in lws_cgi_write_split_stdout_headers().
After lws_cgi_write_split_stdout_headers() gives control to event loop to send out the headers, when it runs again the cgi has been killed.

**Comment by lws-team** [comment]
If in your situation there's no STDOUT for the cgi any more, it's clearly not going to do anything useful after meeting that situation.  It shouldn't crash though even if it's unexpected, the patch just deals with that, but it's not the root cause.

What is your cgi actually doing?  Basically the CGI is aimed at making stuff like mailman work, you're using it when touching mailman areas on libwebsockets.org, eg

https://libwebsockets.org/mailman/listinfo/libwebsockets

> lws_cgi_remove_and_kill() get called, when payload has not been processed in 

There's a race between noticing the subprocess has exited and dealing with pending data... does it make the problem go away if your subprocess hangs around a little while after finishing?

**Comment by zhiwenzheng** [comment]
Yes. sleep(3) after flush the output in cgi app can avoid the problem.
The cgi app does a normal query, return json string in response.
Is doing the same check in lws_cgi_kill_terminated() appropriate to close the race?
'''
         /* wait for stdout to be drained */
1164         if (cgi->content_length > cgi->content_length_seen)
1165             continue;
'''

**Comment by lws-team** [comment]
It's the right idea but it's not so simple I think... waitpid() we call in there is a one-time,systemwide  thing and he will just report what he sees about dead child processes.  Once he reaps the child process and reports it, he is done and won't report it again.  If we ignore it one time due to seeing buffered content there, we won't get asked again.  waitpid() is the best way, because it's the only way we know that it's really our child process that went away.  Trying to match on extant pids leaves it open that we're looking at a completely different process that reused the pid.

Content length is not enough either since it's common with CGI they don't issue any content length, just do chunked transfer-encoding until they find that they finished... they don't know how much they will emit before they do it either.

Really solving it I think means keeping an lws_dll2 list in the context of child pids that came up in waitpid() but we gave extra time to.  And in `lws_cgi_kill_terminated()`, go through the list and do the deferred close processing for any process that hung around for more than a few seconds.

**Comment by zhiwenzheng** [comment]
I need sometime to read the code to understand your idea.
I want to notify you some other problems I encountered when I try cgi:

1)  SCRIPT_PATH should be SCRIPT_NAME (rfc3875)
```
403     env_array[n++] = p;
404     p += lws_snprintf(p, end - p, "SCRIPT_PATH=%s", exec_array[0]) + 1;
```

2) do setenv between vfork() and exec() may not be alright, 
```
498 #if !defined(LWS_HAVE_VFORK) || !defined(LWS_HAVE_EXECVPE)
 499     for (m = 0; m < n; m++) {
 500         p = strchr(env_array[m], '=');
 501         *p++ = '\0';
 502         setenv(env_array[m], p, 1);
 503     }
 504     execvp(exec_array[0], (char * const *)&exec_array[0]);
 505 #else
```
I run the code in illumos, which has vfork() and not execvpe, doing setenv() after vfork()
does not pass the env right. Can we require the filename be an absolute path, and use execve() in all unix platform?

3) break in line 331 may cause line 333 never has a chance to run,
which can stop the cgi server sending the ending chunk when chunked
transfer is used.
```
326                 wsi->reason_bf &= ~LWS_CB_REASON_AUX_BF__CGI;
327 
328             if (wsi->http.cgi && wsi->http.cgi->cgi_transaction_over)
329                 return -1;
330             break;
331         }
332 
333         if (wsi->reason_bf & LWS_CB_REASON_AUX_BF__CGI_CHUNK_END) {
334             if (!wsi->http2_substream) {
335                 memcpy(buf + LWS_PRE, "0\x0d\x0a\x0d\x0a", 5);
336                 lwsl_debug("writing chunk term and exiting\n");
337                 n = lws_write(wsi, (unsigned char *)buf +
338                            LWS_PRE, 5, LWS_WRITE_HTTP);
339             } else
```

**Comment by lws-team** [comment]
>  SCRIPT_PATH should be SCRIPT_NAME

If you google around, both are in common use... it's like that because one of the cgis I used this with, mailman or cgit, wanted it.  I think it's OK to add SCRIPT_NAME as well.  But when I look around, and at the RFC, it's really poorly specified what they contain.  Basically apache did something back in the day and they wrote the RFC around that.  Eg from Wikipedia https://en.wikipedia.org/wiki/Common_Gateway_Interface#Example

```
SCRIPT_FILENAME="C:/Program Files (x86)/Apache Software Foundation/Apache2.4/cgi-bin/printenv.pl"
SCRIPT_NAME="/cgi-bin/printenv.pl"
```
So "/cgi-bin/..." is a virtual path actually.  There's no basis for making that kind of decision with what lws has to hand.

> vfork

vfork has some significant advantages in cost of spawning the cgi, and it seems to work as it is on Linux.... at least that's what I'm using.  If it gets different results on your kernel, using the other "compatibility" path (by `#if defined(__illumos__)` or whatever) is fine.  But I don't want to downgrade Linux build based on problems that only appear on non-linux platform.

> break

The problem is that the call to `lws_cgi_write_split_stdout_headers()` earlier may have exhausted the writeability on the onward socket.  Recent lws (including v3.2) uses lws_buflist to store data that might not be writeable in the case of double writes, so it should be OK to remove this.  But then it becomes part of the race-terminated-child-process problem (`wsi->buflist_out` is NULL if it has no pending output data in the buffer).

**Comment by zhiwenzheng** [comment]
>
>vfork

I am not suggesting not use vfork, instead not use setenv after vfork.
Modifying data after vfork should be unsafe on any platform. 
The linux man page says it is undefined behavior.
>
>Standard description
       (From  POSIX.1) The vfork() function has the same effect as fork(2), except that the behavior is undefined if the process created by vfork() either modifies any data other than a variable of  type  pid_t used to store the return value from vfork(), or returns from the function in which vfork() was called, or calls any other function before successfully calling _exit(2) or one of the exec(3) family of functions.

It works fine on linux might because it never hits this code path, since vfork and execvpe are both available, execvpe are always used to pass the envs. But on other platforms (e.g. freebsd, macos, solaris) execvpe may be unavailable, in those cases, I think execve is preferred over setenv.

**Comment by lws-team** [comment]
Okay we can try it if you want to send a patch.

**Comment by lws-team** [comment]
I pushed this on master Oct 7th... forgot to close this at the time.
